tls/s2n_client_key_exchange.c

```diff
@@ -49,136 +51,90 @@ static int s2n_rsa_client_key_recv(struct s2n_connection *conn)
 
     encrypted.size = s2n_stuffer_data_available(in);
     encrypted.data = s2n_stuffer_raw_read(in, length);
     notnull_check(encrypted.data);
     gt_check(encrypted.size, 0);
 
     /* First: use a random pre-master secret */
-    GUARD(s2n_get_private_random_data(&pms));
+    GUARD(s2n_get_private_random_data(shared_key));
     conn->secure.rsa_premaster_secret[0] = client_protocol_version[0];
     conn->secure.rsa_premaster_secret[1] = client_protocol_version[1];
 
     /* Set rsa_failed to 1 if s2n_pkey_decrypt returns anything other than zero */
-    conn->handshake.rsa_failed = !!s2n_pkey_decrypt(&conn->config->cert_and_key_pairs->private_key, &encrypted, &pms);
+    conn->handshake.rsa_failed = !!s2n_pkey_decrypt(&conn->config->cert_and_key_pairs->private_key, &encrypted, shared_key);
 
     /* Set rsa_failed to 1, if it isn't already, if the protocol version isn't what we expect */
-    conn->handshake.rsa_failed |= !s2n_constant_time_equals(client_protocol_version, pms.data, S2N_TLS_PROTOCOL_VERSION_LEN);
-
-    /* Turn the pre-master secret into a master secret */
-    GUARD(s2n_prf_master_secret(conn, &pms));
-
-    /* Erase the pre-master secret */
-    GUARD(s2n_blob_zero(&pms));
-
-    /* Expand the keys */
-    GUARD(s2n_prf_key_expansion(conn));
-
-    /* Save the master secret in the cache */
-    if (s2n_allowed_to_cache_connection(conn)) {
-        GUARD(s2n_store_to_cache(conn));
-    }
-
+    conn->handshake.rsa_failed |= !s2n_constant_time_equals(client_protocol_version, shared_key->data, S2N_TLS_PROTOCOL_VERSION_LEN);
     return 0;
 }
 
-static int s2n_dhe_client_key_recv(struct s2n_connection *conn)
+int s2n_dhe_client_key_recv(struct s2n_connection *conn, struct s2n_blob *shared_key)
 {
     struct s2n_stuffer *in = &conn->handshake.io;
-    struct s2n_blob shared_key = {0};
 
     /* Get the shared key */
-    if (conn->secure.cipher_suite->key_exchange_alg->flags & S2N_KEY_EXCHANGE_ECC) {
-        GUARD(s2n_ecc_compute_shared_secret_as_server(&conn->secure.server_ecc_params, in, &shared_key));
-    } else {
-        GUARD(s2n_dh_compute_shared_secret_as_server(&conn->secure.server_dh_params, in, &shared_key));
-    }
-
-    /* Turn the pre-master secret into a master secret */
-    GUARD(s2n_prf_master_secret(conn, &shared_key));
-
-    /* Erase the pre-master secret */
-    GUARD(s2n_blob_zero(&shared_key));
-    GUARD(s2n_free(&shared_key));
-
-    /* Expand the keys */
-    GUARD(s2n_prf_key_expansion(conn));
+    GUARD(s2n_dh_compute_shared_secret_as_server(&conn->secure.server_dh_params, in, shared_key));
+    /* We don't need the server params any more */
+    GUARD(s2n_dh_params_free(&conn->secure.server_dh_params));
+    return 0;
+}
 
-    /* Save the master secret in the cache */
-    if (s2n_allowed_to_cache_connection(conn)) {
-        GUARD(s2n_store_to_cache(conn));
-    }
+int s2n_ecdhe_client_key_recv(struct s2n_connection *conn, struct s2n_blob *shared_key)
```

Comment:
Looking on the s2n_ecdhe/dh/rsa_client_key_recv/send functions you have 6 functions 3+3 all do the same 
client: 1) calc_shared_as_client; 2) free_params
server: 1) calc_shared_as_server; 2) free_params

You already entered 4 functions variable into the structure... why not adding another three and combine these functions into only 2 (instead of 6)? later when you will have hybrid it will remain 2 instead of 6+hybrid....?